sql/catalyst/src/main/java/org/apache/spark/sql/connector/CustomMetric.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.read.Scan;
+
+/**
+ * A custom metric. This is a logical representation of a metric reported by data sources during
+ * read path. Data sources can report supported metric list by {@link Scan} to Spark in query
+ * planning. During query execution, Spark will collect the metrics per partition by
+ * {@link PartitionReader} and combine metrics from partitions to the final result. How Spark
+ * combines metrics depends on the metric type. For streaming query, Spark will collect and combine
+ * metrics for a final result per micro batch.
+ *
+ * The metrics will be gathered during query execution back to the driver and then combined. The
+ * final result will be shown up in the physical operator in Spark UI.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public interface CustomMetric {
+    /**
+     * Returns the name of custom metric.
+     */
+    String name();
+
+    /**
+     * Returns the description of custom metric.
+     */
+    String description();
+
+    /**
+     * Supported metric type. The metric types must be supported by Spark SQL internal metrics.
```

Comment:
Spark SQL internal metrics always return a long value. Do we still need the `LongMetric` interface if metrics types are limited to SQL internal metrics?

Reply:
For now a `CustomMetric` which returns long value should be enough. Previously there is a suggestion to name it just `LongMetric` because it always returns long value. But it sounds too concrete.

So I think is it good to leave some flexibility here? Then I extract `LongMetric` out as a separate interface which returns long value.

I'm fine to just return long in `CustomMetric` (and maybe name it to `LongMetric`).

Reply:
I'll be +1 on returning the long value directly in `CustomMetric`. I don't have a strong opinion about the naming.

Reply:
Okay, let me move it back. See if others have different thoughts.

Reply:
> The metric types must be supported by Spark SQL internal metrics.

It sounds weird that we require users to understand internal private APIs when using a public API.

Reply:
"The metric types must be supported by Spark SQL internal metrics." is more for the developers trying to add new metrics here, instead of using the API. I can remove it if it sounds inappropriate.

Reply:
+1 for removing it. I think users just need to know the below line: `SUM: Spark sums up metrics from partitions as the final result.`.